### PR TITLE
use DeployUtils lib for deployments in deploy script

### PR DESCRIPTION
**Pull request by AmadiMichael**
This PR replaces the use of the `_deploy` function in `Deploy.s.sol` with deploy helper functions in `libraries/DeployUtils.sol`

### CODE REVIEW

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol [hunk 1]
```diff
@@ -1596,17 +1630,19 @@ contract Deploy is Deployer {
                             _params.weth,
                             IAnchorStateRegistry(mustGetAddress("AnchorStateRegistryProxy")),
                             cfg.l2ChainID()
-                        )
-                    )
+                        ),
```
Comment (tynes) [inline packages/contracts-bedrock/scripts/deploy/Deploy.s.sol]: Are you on the version of forge defined in `versions.json`? Looks like there are some listing things that are off

Comment (AmadiMichael) [reply packages/contracts-bedrock/scripts/deploy/Deploy.s.sol]: Yeah i am,. Is the issue still present?

File: packages/contracts-bedrock/scripts/libraries/DeployUtils.sol [hunk 2]
```diff
@@ -198,6 +199,12 @@ library DeployUtils {
         return address(uint160(uint256(keccak256(abi.encode(_sender, _identifier)))));
     }
 
+    /// @notice Strips the first 4 bytes of `_data` and returns the remaining bytes
+    ///         If `_data` is not greater than 4 bytes, it returns empty bytes type.
+    function encodeConstructor(bytes memory _data) internal pure returns (bytes memory _encodedData) {
+        _encodedData = _data.length > 4 ? Bytes.slice(_data, 4) : new bytes(0);
```
Comment (smartcontracts) [inline packages/contracts-bedrock/scripts/libraries/DeployUtils.sol]: As a sanity check can you also check that the input is at least 4 bytes long?

File: packages/contracts-bedrock/scripts/libraries/DeployUtils.sol [hunk 3]
```diff
@@ -198,6 +199,12 @@ library DeployUtils {
         return address(uint160(uint256(keccak256(abi.encode(_sender, _identifier)))));
     }
 
+    /// @notice Strips the first 4 bytes of `_data` and returns the remaining bytes
+    ///         If `_data` is not greater than 4 bytes, it returns empty bytes type.
```
Comment (smartcontracts) [inline packages/contracts-bedrock/scripts/libraries/DeployUtils.sol:203]: Needs params/returns comments

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol [hunk 4]
```diff
@@ -1578,49 +1648,64 @@ contract Deploy is Deployer {
         }
 
         uint32 rawGameType = GameType.unwrap(_params.gameType);
+        FaultDisputeGameParams memory _params_ = _params;
```
Comment (smartcontracts) [inline packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:1699]: Why is this required?

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol [hunk 5]
```diff
@@ -888,7 +950,9 @@ contract Deploy is Deployer {
 
     /// @notice Deploy the L1ERC721Bridge
     function deployL1ERC721Bridge() public broadcast returns (address addr_) {
-        IL1ERC721Bridge bridge = IL1ERC721Bridge(_deploy("L1ERC721Bridge", hex""));
+        IL1ERC721Bridge bridge = IL1ERC721Bridge(
+            DeployUtils.create2AndSave({ _save: this, _name: "L1ERC721Bridge", _args: hex"", _salt: _implSalt() })
+        );
```
Comment (smartcontracts) [inline packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:1001]: Let's use constructor encoding instead of hex""

Comment (smartcontracts) [reply packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:1001]: Applies in a number of locations

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol [hunk 6]
```diff
@@ -1578,49 +1648,64 @@ contract Deploy is Deployer {
...
+                        _nick: string.concat("FaultDisputeGame_", vm.toString(rawGameType)),
+                        _args: DeployUtils.encodeConstructor(
+                            abi.encodeCall(
+                                IFaultDisputeGame.__constructor__,
+                                (
+                                    _params_.gameType,
+                                    _params_.absolutePrestate,
+                                    _params_.maxGameDepth,
+                                    cfg.faultGameSplitDepth(),
+                                    _unsafeCastUint256ToDuartion(cfg.faultGameClockExtension()),
```
Comment (smartcontracts) [inline packages/contracts-bedrock/scripts/deploy/Deploy.s.sol]: `Duration.wrap(uint64(cfg.faultGameClockExtension()))`

Can use this instead

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol [hunk 7]
```diff
@@ -849,19 +893,35 @@ contract Deploy is Deployer {
 
     /// @notice Deploy the AnchorStateRegistry
     function deployAnchorStateRegistry() public broadcast returns (address addr_) {
-        IAnchorStateRegistry anchorStateRegistry =
-            IAnchorStateRegistry(_deploy("AnchorStateRegistry", abi.encode(mustGetAddress("DisputeGameFactoryProxy"))));
+        IAnchorStateRegistry anchorStateRegistry = IAnchorStateRegistry(
+            DeployUtils.create2AndSave({
+                _save: this,
+                _name: "AnchorStateRegistry",
```
Comment (smartcontracts) [inline packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:931]: Minor nit but for aesthetics let's put the `_salt` parameter underneath the `_save` parameter, will make the code look a bit nicer

Comment (smartcontracts) [reply packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:931]: Worth applying everywhere in this file

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol (hunk 4 above)
Comment (AmadiMichael) [reply packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:1699]: ran into stack too deep errors

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol (hunk 6 above)
Comment (AmadiMichael) [reply packages/contracts-bedrock/scripts/deploy/Deploy.s.sol]: yeah definitely

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol (hunk 4 above)
Comment (smartcontracts) [reply packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:1699]: Ah weird, can you leave a comment to that effect?

File: packages/contracts-bedrock/scripts/checks/check-interfaces.sh [hunk 8]
```diff
@@ -208,6 +208,29 @@ for interface_file in $JSON_FILES; do
     normalized_interface_abi=$(normalize_abi "$interface_abi")
     normalized_contract_abi=$(normalize_abi "$contract_abi")
 
+    # Check if the contract ABI has no constructor but the interface is missing __constructor__
+    contract_has_constructor=$(echo "$normalized_contract_abi" | jq 'any(.[]; .type == "constructor")')
+    interface_has_pseudo_constructor=$(echo "$normalized_interface_abi" | jq 'any(.[]; .type == "constructor")')
+
+    # If any contract has no constructor and its corresponding interface also does not have one, flag it as a detected issue
+    if [ "$contract_has_constructor" = false ] && [ "$interface_has_pseudo_constructor" = false ]; then
```
Comment (smartcontracts) [inline packages/contracts-bedrock/scripts/checks/check-interfaces.sh]: Maybe we can explicitly check for a `$contract_has_constructor = false && $interface_has_empty_pseudo_constructor`

**Review by smartcontracts — APPROVED** [review]
